File: 121. Best Time to Buy and Sell Stock/maxProfit.java
class Solution {
    //Dynamic programming. Using OPT[i] refers to the max profit we can get from first i days.
    /*OPT[i] = OPT[i-1] do nothing in day i or buy ticket in day i.
     OPT[i] = price[i] - smallest_value_from_first_i-1_days.
     */
    public int maxProfit(int[] prices) {
        //one exception: no day
        if(prices.length == 0)  return 0;
        int[] OPT = new int[prices.length];
        int smallest = prices[0];
        for(int i = 1; i < prices.length; i++){
            OPT[i] = Math.max(OPT[i-1], prices[i]-smallest);
            smallest = Math.min(smallest, prices[i]);
        }
        return OPT[OPT.length-1];
    }
}

File: 121. Best Time to Buy and Sell Stock/maxProfit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        # min
        # max: max(cur - min, last)
        min_val = float('inf')
        max_val = 0
        for cur in prices:
            min_val = min(cur, min_val)
            max_val = max(cur - min_val, max_val)
        return max_val